pennylane/pulse/transmon.py

```diff
@@ -0,0 +1,206 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""This module contains the classes/functions specific for simulation of superconducting transmon hardware systems"""
+from dataclasses import dataclass
+from typing import Callable, List, Union
+
+import pennylane as qml
+from pennylane.pulse import HardwareHamiltonian
+from pennylane.typing import TensorLike
+
+
+# TODO ladder operators once there is qudit support
+# pylint: disable=unused-argument
+def a(wire, d=2):
+    """creation operator"""
+    return qml.s_prod(0.5, qml.PauliX(wire)) + qml.s_prod(0.5j, qml.PauliY(wire))
+
+
+def ad(wire, d=2):
+    """annihilation operator"""
+    return qml.s_prod(0.5, qml.PauliX(wire)) + qml.s_prod(-0.5j, qml.PauliY(wire))
+
+
+# pylint: disable=too-many-arguments
+def transmon_interaction(
+    omega: Union[float, list],
+    connections: list,
+    g: Union[float, list],
+    anharmonicity=None,
+    wires=None,
+    d=2,
+):
+    r"""Returns a :class:`ParametrizedHamiltonian` representing the cQED Hamiltonian of a superconducting transmon system.
+
+    The Hamiltonian is given by
+
+    .. math::
+
+        H = \sum__{q\in \text{wires}} \omega_q a^\dagger_q a_q
+        + \sum_{(i, j) \in \mathcal{C}} g_{ij} \left(a^\dagger_i a_j + a_j^\dagger a_i \right)
+        + \sum_{q\in \text{wires}} \alpha_q a^\dagger_q a^\dagger_q a_q a_q
+
+    where :math:`[a^\dagger_p, a_q] = i \delta_{pq}` are bosonic creation and annihilation operators.
+    The first term describes the dressed qubit frequencies :math:`\omega_q`, the second term their
+    coupling :math:`g_{ij}` and the last the anharmonicity :math:`\alpha_q`, which all can vary for
+    different qubits. In practice, the bosonic operators are restricted to a finite dimension of the
+    local Hilbert space (default ``d=2`` corresponds to qubits).
+    In that case, the anharmonicity is set to :math:`\alpha=0` and ignored.
+
+    The values of :math:`\omega \pm \alpha` are typically around :math:`(5 \pm 0.3) \times 2\pi \text{GHz}` and it is common
+    for different qubits to be out of tune with different energy gaps. The coupling strength
+    :math:`g` typically varies betwewen :math:`[0.001, 0.1] \times 2\pi \text{GHz}`. For some example parameters,
+    see e.g. `arXiv:1804.04073 <https://arxiv.org/abs/1804.04073>`_,
+    `arXiv:2203.06818 <https://arxiv.org/abs/2203.06818>`_, or `arXiv:2210.15812 <https://arxiv.org/abs/2210.15812>`_.
+
+    .. note:: Currently only supporting ``d=2`` with qudit support planned in the future.
+
+    .. seealso::
+
+        :func:`~.drive`
+
+    Args:
+        omega (Union[float, list[float]]): List of dressed qubit frequencies in GHz.
+            When passing a single float all qubits are assumed to have that same frequency.
```

Comment:
```suggestion
```
I would be inclined to remove this option and say that each qubit frequency should be specified individually. If they should all be identical, pass `[4 for i in range(5)]` instead of `4`. But this depends what you think of my suggestion for the logic determining wires.

Reply:
:+1: I feel like `wires` should not be optional, is there any other case in PL where `wires` is optional?

Reply:
No super strong feelings about this. I feel like it can be a convenience to have that option for writing tight code but maybe it is indeed better to be specific, less room for bugs 👍

Reply:
But if we make `wires` mandatory then we can also keep the option of passing a single float for omega no?

Reply:
So I made wires non-optional and kept the option to pass a single float for omega which I find handy and doesnt cause problems here

I think the only thing that might be a bit weird is that in principle I can pass `connections = [[100, 101], [102, 103]]` and `wires=[0, 1, 2, 3]` as they are treated completely separately. No problem in principle, and I wouldnt want to be too restrictive here, but I added warning when the connections are not contained in wires.